### ISSUE TITLE
Nested page component not re-rendered when data is updated

### ISSUE THREAD

**Issue by PrimozRome** [nuxt/nuxt.js#1754]
I have dynamic nested route `/pages/users/_id.vue`. I use `asyncData` to fetch the data from server. 

```
  async asyncData ({ app, params }) {
    try {
      let { data } = await app.$axios.get(`/users/${params.id}`)
      console.log('data arrived')
      return { user: data.data }
    } catch (e) {
      console.log(e)
    }
  },
```

```
template>
  <div>
      <h3 class="semi-bold">
        {{ user.firstname) }} {{ user.lastname }}
      </h3>
   </div>
</template>
```

I navigate to `localhost:3000/users/50` the page is rendered correctly. Then I navigate to other user page `localhost:3000/users/567` the asyncData is executed and new user returned but the view is not re-rendered...

Is this a bug ?

**Comment by PrimozRome** [comment]
Hmm if I restart server and re-run with `npm run dev` then its working... but only for a while and then again stops re-rendering at some time. Very strange...

**Comment by tillkolter** [comment]
By navigating, do you mean going directly from one page to another through a link?

I was facing this problem and found that if you stay on the same route with only the parameters changing, the page component itself is not recreated and with that the async page rendering is not triggered. The pattern to use is to introduce a watcher on the `$route` object.

```
watch: {
      $route ({params, query}) {
        this.$axios.get(`/users/${params.id}`).then(....)
    }
}
```

I am still running on an older version, so this might have changed, but I think this router/component behaviour is the basic Vue behaviour, so my suggestion should still work the same.

**Comment by PrimozRome** [comment]
> By navigating, do you mean going directly from one page to another through a link?

YES indeed. It's very strange behaviour because as I said it works for a while and then stops working at some time.

I will try your suggestion, but I think this should be fixed asap as `asyncData` is intended for this anyway. And it is triggered and it updates the component data so I think re-render should occur always.

**Comment by tillkolter** [comment]
@PrimozRome When I first came across this problem, I thought the same. Why the heck is the component object reused and not constructed from scratch? It does not only introduce evil code snippet twins for the data fetches, but you also have to keep in mind that the whole component state has to be reset inside the watcher handle, in case other data properties are involved.

But if you think twice you could also argue the other way around. Keeping the already rendered object alive and just update it, makes the whole thing faster. Also keep in mind that the purpose of using `asyncData` is not a pure matter of fashion, but to make a page source code readable if called by a web crawler which is not acting like a user navigating with a browser but calls each page separately. If not for the google bots and social shares I would always prefer an async version, because you can play loading animations during requests and give a better overall experience to the user.

**Comment by tjbenton** [comment]
I'm having this same problem. This seems like a really bad bug to me. When the I use the `nuxt-link` to change the url to a different page it only changes to the page correctly 1 out of 10 times I try it, and the other 9 times I have to do a full reload of the page to get it to display the correct page.

I'm new to Vue and I'm not sure how to get the page to render correctly when the url changes.

**Comment by tillkolter** [comment]
@tjbenton Did you try what i was suggesting?

**Comment by alexchopin** [comment]
Hi, 
You can also create a computed property to bind changes on your data.
```js
  async asyncData ({ app, params }) {
    try {
      let { data } = await app.$axios.get(`/users/${params.id}`)
      console.log('data arrived')
      return { user: data.data }
    } catch (e) {
      console.log(e)
    }
  },
  computed: {
    currentUser () {
      return this.user
    }
  }
```

**Comment by tjbenton** [comment]
@tillkolter @alexchopin I have tried both of these approaches and I haven't had any luck with either of them. I've also tried what they suggest in the docs for [data fetching](https://router.vuejs.org/en/advanced/data-fetching.html). 

Here's my fake data function that I used in the examples below
<details>
<summary>Fake data function</summary>

```js
// fake get data function to simulate api call
function getData(type) {
  const fake_data = {
    device: {
      data_points: [ { data_id: 'device 1', users: 20 }, { data_id: 'device 2', users: 40 }, { data_id: 'device 3', users: 60 }, { data_id: 'device 4', users: 80 } ],
      data_message: 'device data message',
    },
    geo: {
      data_points: [ { data_id: 'geo 1', users: 200 }, { data_id: 'geo 2', users: 400 }, { data_id: 'geo 3', users: 600 }, { data_id: 'geo 4', users: 800 } ],
      data_message: 'geo data message',
    },
  };
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        data_points: fake_data[type].data_points,
        data_message: fake_data[type].data_message,
      });
    }, 200);
  });
}
```
</details>

<br>

### Method 1 

In the first method I tried using computed property to bind the changes to the data. 

<details>
<summary>Method 1 Code</summary>

```vue
<template lang="pug">
  md-whiteframe(style="padding: 1em")
    h1(style="line-height: 2.45em;") {{dataMessage}}
    md-toolbar
      md-button(class="md-raised md-primary"): nuxt-link(to="/datasets/device") device
      md-button(class="md-raised md-primary"): nuxt-link(to="/datasets/geo") geo

    md-whiteframe
      h3 Data Points
      pre(v-highlightjs="JSON.stringify(dataPoints, null, 2)"): code.json

</template>

<script>
export default {
  async asyncData({ params, store }) {
    console.log('asyncData pages/datasets/_type.vue');
    const data = await getData(params.type);
    store.commit('setTitle', params.type);
    console.log('data:', data);
    return data;
  },

  computed: {
    dataMessage() {
      return this.data_message;
    },
    dataPoints() {
      return this.data_points;
    },
  },
};
</script>
```
</details>

<br>

![http://recordit.co/9KnCSACipp](http://g.recordit.co/9KnCSACipp.gif)

In the result of running this code there's a couple weird things that I notice

  - The place where I used a store commit updates correctly when the url changes but the rest of the data doesn't update as expected.
  - When the page first loads the vue developer tools show that there's a `WebPagesDatasets_type.vue` component that's inside of the main `Nuxt` component, and the data is correct. However when I click on `geo` and the url changes via `nuxt-link` the `WebPagesDatasets_type.vue` component disappears and you can't see if the data has been updated in the vue developer tools.
  - The console shows that the data being set is correct, but nothing actually updates on the page.


<br>

### Method 2

In method 2 I tried to use the [data fetching](https://router.vuejs.org/en/advanced/data-fetching.html) method that the router docs give an example of.


<details>
<summary>Method 2 Code</summary>

```vue
<template lang="pug">
  md-whiteframe(style="padding: 1em")
    h1(style="line-height: 2.45em;") {{dataMessage}}
    md-toolbar
      md-button(class="md-raised md-primary"): nuxt-link(to="/datasets/device") device
      md-button(class="md-raised md-primary"): nuxt-link(to="/datasets/geo") geo

    md-whiteframe
      h3 Data Points
      pre(v-highlightjs="JSON.stringify(dataPoints, null, 2)"): code.json

</template>

<script>
export default {
  data() {
    return {
      data_points: [ { data_id: 'initial data id', users: 0 } ],
      data_message: 'initial data message',
    };
  },

  async beforeRouteEnter(to, from, next) {
    console.log('beforeRouteEnter:');
    const data = await getData(to.params.type);
    next((vm) => vm.setData('beforeRouteEnter:', data));
  },

  async beforeRouteUpdate(to, from, next) {
    console.log('beforeRouteUpdate:');
    const data = await getData(to.params.type);
    this.setData('beforeRouteUpdate', data);
    next();
  },

  methods: {
    setData(called_from, data = {}) {
      console.log(`${called_from} setData:`);
      this.$store.commit('setTitle', this.$route.params.type);
      this.data_points = data.data_points;
      this.data_message = data.data_message;
      console.log('this.$data:', this.$data);
    },
  },

  computed: {
    dataMessage() {
      return this.data_message;
    },
    dataPoints() {
      return this.data_points;
    },
  },
};
</script>
```
</details>

<br>

![http://recordit.co/9A9giMu2jH](http://g.recordit.co/9A9giMu2jH.gif)


After running this code here's a few things I notice that were different from method 1.

  - When `geo` was clicked  the page changed but the store title didn't update like it did in method 1
  - When `geo` was clicked the data changed but it changed to be the initial data which it shouldn't do 
  - When `geo` was clicked the `Nuxt` component still had a child component unlike method 1, but the data was still incorrect.
  - Not shown in the gif, but when I saved a file in my project the page would refresh and show the initial data. Not the correct data for the current page.

<br>

### Method 3

In method 3 I tried to use the watch method of the [data fetching](https://router.vuejs.org/en/advanced/data-fetching.html) method in the docs.

<details>
<summary>Method 3 Code</summary>

```vue
<template lang="pug">
  md-whiteframe(style="padding: 1em")
    h1(style="line-height: 2.45em;") {{dataMessage}}
    md-toolbar
      md-button(class="md-raised md-primary"): nuxt-link(to="/datasets/device") device
      md-button(class="md-raised md-primary"): nuxt-link(to="/datasets/geo") geo

    md-whiteframe
      h3 Data Points
      pre(v-highlightjs="JSON.stringify(dataPoints, null, 2)"): code.json

</template>

<script>
export default {
  data() {
    return {
      data_points: [ { data_id: 'initial data id', users: 0 } ],
      data_message: 'initial data message',
    };
  },

  created() {
    this.fetchData('created');
  },

  watch: {
    // call again the method if the route changes
    $route: 'fetchData',
  },

  methods: {
    async fetchData(called_from = 'watch') {
      console.log(`${called_from} fetchData:`);
      const { type } = this.$route.params;
      this.$store.commit('setTitle', type);
      const data = await getData(type);
      this.data_points = data.data_points;
      this.data_message = data.data_message;
      console.log('this.$data:', this.$data);
    },
  },

  computed: {
    dataMessage() {
      return this.data_message;
    },
    dataPoints() {
      return this.data_points;
    },
  },
};
</script>
```
</details>

<br>

![http://recordit.co/1RGbhggttv](http://g.recordit.co/1RGbhggttv.gif)

This method did the least out of all of them. Only 2 things changed on the page.

  - The url changed to be `geo`
  - The `Nuxt` components child `WebPagesDatasets_type.vue` was removed when `geo` was clicked.
  - No more logs happened when `geo` was clicked 



<br>


Since my current project only has 4 different possibilities for the type param I just duplicated the same file 4 times so that I could get it finished but it would be nice to go back and update this to be dynamic.

**Comment by tjbenton** [comment]
After putting all that together to show the issue I was running into I figured out that the entire issue was due to [vue-material](https://github.com/vuematerial/vue-material/issues/243) not currently supporting SSR which causes binding issues and data doesn't get updated correctly as it should. They're about to release a [beta](https://github.com/vuematerial/vue-material/issues/1081) version that has a bunch of fixes and SSR is at the top of it's list. 

After moving to [vuetifyjs](https://vuetifyjs.com/) I haven't experienced the same problems with dynamic routes not working correctly.
